pkg/pbq/pbqmanager.go

```diff
@@ -0,0 +1,164 @@
+	}
+	return pbqManager, nil
+}
+
+// ListPartitions returns all the pbq instances
+func (m *Manager) ListPartitions() []*PBQ {
+
+	pbqList := make([]*PBQ, len(m.pbqMap))
+	i := 0
+	for _, val := range m.pbqMap {
+		pbqList[i] = val
+		i++
+	}
+	return pbqList
+}
+
+// GetPBQ returns pbq for the given partitionID, if createIfMissing is set to true
+// it creates and return a new pbq instance
+func (m *Manager) GetPBQ(ctx context.Context, partitionID string, createIfMissing bool) (*PBQ, bool, error) {
+
+	pbqInstance, ok := m.pbqMap[partitionID]
+	if ok {
+		return pbqInstance, false, nil
+	}
+
+	// no match and nothing to create
+	if !ok && !createIfMissing {
+		m.log.Info("requested pbq is not created", zap.Any("partitionID", partitionID))
+		return nil, false, NonExistentPBQErr
+	}
+
+	// no match and create if missing
+	var persistentStore store.Store
+	var err error
+
+	switch m.options.PbqStoreType() {
+	case dfv1.NoOpType:
+		persistentStore, _ = noop.NewPBQNoOpStore()
+	case dfv1.InMemoryType:
+		persistentStore, err = memory.NewMemoryStore(ctx, partitionID, m.options)
+		if err != nil {
+			m.log.Errorw("Error while creating persistent store", zap.Any("partitionID", partitionID), zap.Any("store type", m.options.PbqStoreType()), zap.Error(err))
+			return nil, true, err
+		}
+	case dfv1.FileSystemType:
+		return nil, true, errors.New("not implemented")
+	}
+	pbq, err := NewPBQ(ctx, partitionID, persistentStore, m, m.options)
+	if err != nil {
+		m.log.Errorw("Error while creating PBQ", zap.Any("Partition ID", partitionID), zap.Error(err))
+		return nil, true, err
+	}
+	m.Register(partitionID, pbq)
+	return pbq, true, nil
+}
+
+// StartUp creates list of PBQ instances using the persistent store
+func (m *Manager) StartUp(ctx context.Context) error {
+
+	switch m.options.PbqStoreType() {
+	case dfv1.NoOpType:
+		return nil
+	case dfv1.InMemoryType:
+		return nil
+	case dfv1.FileSystemType:
+		return errors.New("not implemented")
+	}
+
+	for _, value := range m.pbqMap {
+		// set replay flag so that it replays messages from the store during startup
+		value.SetIsReplaying(ctx, true)
+	}
+	return nil
+}
+
+// ShutDown for clean shut down, flushes pending messages to store and closes the store
+func (m *Manager) ShutDown(ctx context.Context) {
+	// iterate through the map of pbq
+	// close all the pbq
+	var wg sync.WaitGroup
+
+	var PBQCloseBackOff = wait.Backoff{
+		Steps:    5,
+		Duration: 1 * time.Second,
+		Factor:   1.5,
+		Jitter:   0.1,
+		Cap:      5 * time.Second,
+	}
+
+	for _, v := range m.pbqMap {
+		wg.Add(1)
+		go func(q *PBQ) {
+			defer wg.Done()
+			var closeErr error
+			var attempt int
+			closeErr = wait.ExponentialBackoffWithContext(ctx, PBQCloseBackOff, func() (done bool, err error) {
+				closeErr = q.CloseWriter()
+				if closeErr != nil {
+					attempt += 1
+					m.log.Errorw("Failed to close pbq, retrying", zap.Any("attempt", attempt), zap.Any("partitionID", q.partitionID), zap.Error(closeErr))
+					return false, err
```

Comment:
When you return error, the retry ends.

Reply:
yeah realised it now, I am returning nil now